ss/rocksdb/db.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	return copyAndFreeSlice(slice), nil
+}
+
+func (db *Database) ApplyChangeset(version uint64, cs *iavl.Changeset) error {
+	b := NewBatch(db, version)
+
+	for _, kvPair := range cs.Pairs {
+		if kvPair.Value == nil {
+			if err := b.Delete(kvPair.StoreKey, kvPair.Key); err != nil {
+				return err
+			}
+		} else {
+			if err := b.Set(kvPair.StoreKey, kvPair.Key, kvPair.Value); err != nil {
+				return err
+			}
+		}
+	}
+
+	return b.Write()
+}
+
+// Prune attempts to prune all versions up to and including the provided version.
+// This is done internally by updating the full_history_ts_low RocksDB value on
+// the column families, s.t. all versions less than full_history_ts_low will be
+// dropped.
+//
+// Note, this does NOT incur an immediate full compaction, i.e. this performs a
+// lazy prune. Future compactions will honor the increased full_history_ts_low
+// and trim history when possible.
+func (db *Database) Prune(version uint64) error {
+	tsLow := version + 1 // we increment by 1 to include the provided version
+
+	var ts [TimestampSize]byte
+	binary.LittleEndian.PutUint64(ts[:], tsLow)
+
+	if err := db.storage.IncreaseFullHistoryTsLow(db.cfHandle, ts[:]); err != nil {
+		return fmt.Errorf("failed to update column family full_history_ts_low: %w", err)
+	}
+
+	db.tsLow = tsLow
+	return nil
+}
+
+func (db *Database) Iterator(storeKey string, version uint64, start, end []byte) (types.Iterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, ss.ErrKeyEmpty
+	}
+
+	if start != nil && end != nil && bytes.Compare(start, end) > 0 {
+		return nil, ss.ErrStartAfterEnd
+	}
+
+	prefix := storePrefix(storeKey)
+	start, end = util.IterateWithPrefix(prefix, start, end)
+
+	itr := db.storage.NewIteratorCF(newTSReadOptions(version), db.cfHandle)
+	return NewRocksDBIterator(itr, prefix, start, end, false), nil
+}
+
+func (db *Database) ReverseIterator(storeKey string, version uint64, start, end []byte) (types.Iterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, ss.ErrKeyEmpty
+	}
+
+	if start != nil && end != nil && bytes.Compare(start, end) > 0 {
+		return nil, ss.ErrStartAfterEnd
+	}
+
+	prefix := storePrefix(storeKey)
+	start, end = util.IterateWithPrefix(prefix, start, end)
+
+	itr := db.storage.NewIteratorCF(newTSReadOptions(version), db.cfHandle)
+	return NewRocksDBIterator(itr, prefix, start, end, true), nil
+}
+
+// newTSReadOptions returns ReadOptions used in the RocksDB column family read.
+func newTSReadOptions(version uint64) *grocksdb.ReadOptions {
+	var ts [TimestampSize]byte
+	binary.LittleEndian.PutUint64(ts[:], version)
+
+	readOpts := grocksdb.NewDefaultReadOptions()
+	readOpts.SetTimestamp(ts[:])
+
+	return readOpts
+}
+
+func storePrefix(storeKey string) []byte {
+	return []byte(fmt.Sprintf(StorePrefixTpl, storeKey))
+}
+
+func prependStoreKey(storeKey string, key []byte) []byte {
+	if storeKey == "" {
```

Comment:
NOTE: This is an addition thats needed especially for benchmark, where we don't prepend store keys